sematic/plugins/building/docker_builder_config.py

```diff
@@ -0,0 +1,563 @@
+        """
+        kwargs: Dict[str, Any] = {}
+
+        if other.platform is not None:
+            kwargs["platform"] = other.platform
+
+        if other.requirements is not None:
+            kwargs["requirements"] = other.requirements
+
+        if other.data:
+            kwargs["data"] = (self.data or []) + other.data
+
+        if other.src:
+            kwargs["src"] = (self.src or []) + other.src
+
+        return replace(self, **kwargs)
+
+
+@dataclass
+class ImagePushConfig:
+    """
+    A packaged build image push configuration.
+    """
+
+    registry: str
+    repository: str
+    tag_suffix: Optional[str] = None
+
+    def __post_init__(self):
+        """
+        Validates the contents of this object.
+        """
+        if not self.registry or not self.repository:
+            raise BuildConfigurationError(
+                "When `push` is specified, `registry` and `repository` must be non-empty!"
+            )
+
+    def get_repository_str(self) -> str:
+        """
+        Returns a string that identifies the repository according to the `docker` library
+        naming convention.
+        """
+        return f"{self.registry}/{self.repository}"
+
+    def get_tag(self) -> str:
+        """
+        Returns the effective tag to use for the image.
+        """
+        if not self.tag_suffix:
+            return "default"
+        return f"default_{self.tag_suffix}"
+
+    def merge(self, other: "ImagePushConfig") -> "ImagePushConfig":
+        """
+        Merges this `ImagePushConfig` object with the specified one.
+        """
+        if other.tag_suffix is not None:
+            return other
+        return replace(other, tag_suffix=self.tag_suffix)
+
+
+@dataclass
+class DockerClientConfig:
+    """
+    The Docker Client connection details to use to connect to the Docker Server.
+
+    See https://docker-py.readthedocs.io/en/stable/client.html#client-reference for more
+    details.
+    """
+
+    base_url: str = "unix://var/run/docker.sock"
+    version: str = "auto"
+    timeout: Optional[int] = None
+    tls: bool = False
+    user_agent: Optional[str] = None
+    credstore_env: Optional[Dict[str, str]] = None
+    use_ssh_client: bool = False
+    max_pool_size: Optional[int] = None
+
+    def __post_init__(self):
+        """
+        Validates the contents of this object.
+        """
+        if self.timeout is not None and isinstance(self.timeout, str):
+            self.timeout = int(self.timeout)
+
+        self.tls = as_bool(self.tls)
+
+        if self.credstore_env is not None and isinstance(self.credstore_env, str):
+            self.credstore_env = json.loads(self.credstore_env)
+
+        self.use_ssh_client = as_bool(self.use_ssh_client)
+
+        if self.max_pool_size is not None and isinstance(self.max_pool_size, str):
+            self.max_pool_size = int(self.max_pool_size)
+
+    def merge(self, other: "DockerClientConfig") -> "DockerClientConfig":
```

Comment:
You could probably generalize this and use it across all of these:

```python
T = TypeVar("T")

def _merge(config_a: T, config_b: T) -> T:
    if not isdataclass(config_a, config_b):
        return config_b
    
    replacements = {}
    for name in config_b.__dataclass_fields__.keys():
        replacements[name] = _merge(getattr(name, config_a), getattr(name, config_b))
    return replace(config_a, **replacements)
```

Just a suggestion; not a requirement.